pkg/networkservice/chains/chainstest/builder.go

```diff
@@ -0,0 +1,264 @@
+		Registry: reg,
+		URL:      u,
+	}
+	for i := 0; i < b.nodesCount; i++ {
+		var node = new(Node)
+		mgr, u := b.newNSMgr(ctx, domain.Registry.URL)
+		node.NSMgr = &NSMgrEntry{
+			Nsmgr: mgr,
+			URL:   u,
+		}
+		forwarderName := "cross-nse" + uuid.New().String()
+		forwarder, u := b.newCrossConnectNSE(ctx, forwarderName, node.NSMgr.URL)
+		node.Forwarder = &EndpointEntry{
+			Endpoint: forwarder,
+			URL:      u,
+		}
+		forwarderRegistrationClient := chain.NewNetworkServiceEndpointRegistryClient(
+			interpose_reg.NewNetworkServiceEndpointRegistryClient(),
+			adapter_registry.NetworkServiceEndpointServerToClient(node.NSMgr.NetworkServiceEndpointRegistryServer()),
+		)
+		_, err := forwarderRegistrationClient.Register(context.Background(), &registryapi.NetworkServiceEndpoint{
+			Url:  node.Forwarder.URL.String(),
+			Name: forwarderName,
+		})
+		b.require.NoError(err)
+		domain.Nodes = append(domain.Nodes, node)
+	}
+	supplier := &Supplier{
+		resources:    b.resources,
+		supplyNSC:    b.supplyNSC,
+		supplyNSE:    b.supplyNSE,
+		tokenGenFunc: b.generateTokenFunc,
+		require:      b.require,
+		dialContext:  b.dialContext,
+		ctx:          ctx,
+	}
+	b.resources = nil
+	return domain, supplier
+}
+
+// SetContext sets context for all chains
+func (b *DomainBuilder) SetContext(ctx context.Context) *DomainBuilder {
+	b.ctx = ctx
+	return b
+}
+
+// SetNodesCount sets nodes count
+func (b *DomainBuilder) SetNodesCount(nodesCount int) *DomainBuilder {
+	b.nodesCount = nodesCount
+	return b
+}
+
+// SetTokenGenerateFunc sets function for the token generation
+func (b *DomainBuilder) SetTokenGenerateFunc(f token.GeneratorFunc) *DomainBuilder {
+	b.generateTokenFunc = f
+	return b
+}
+
+// SetRegistrySupplier replaces default memory registry supplier to custom function
+func (b *DomainBuilder) SetRegistrySupplier(f SupplyRegistryFunc) *DomainBuilder {
+	b.supplyRegistry = f
+	return b
+}
+
+// SetForwarderSupplier replaces default dummy forwarder supplier to custom function
+func (b *DomainBuilder) SetForwarderSupplier(f SupplyForwarderFunc) *DomainBuilder {
+	b.supplyForwarder = f
+	return b
+}
+
+// SetNSMgrSupplier replaces default nsmgr supplier to custom function
+func (b *DomainBuilder) SetNSMgrSupplier(f SupplyNSMgrFunc) *DomainBuilder {
+	b.supplyNSMgr = f
+	return b
+}
+
+func (b *DomainBuilder) dialContext(ctx context.Context, u *url.URL) *grpc.ClientConn {
+	conn, err := grpc.DialContext(ctx, grpcutils.URLToTarget(u),
+		grpc.WithInsecure(),
+		grpc.WithBlock(),
+		grpc.WithDefaultCallOptions(grpc.WaitForReady(true)))
+	b.resources = append(b.resources, func() {
+		_ = conn.Close()
+	})
+	b.require.NoError(err, "Can not dial to", u)
+	return conn
+}
+
+func (b *DomainBuilder) newNSMgr(ctx context.Context, registryURL *url.URL) (nsmgr.Nsmgr, *url.URL) {
+	var registryCC *grpc.ClientConn
+	if registryURL != nil {
+		registryCC = b.dialContext(ctx, registryURL)
+	}
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
```

Comment:
Should we use just ":0"?

Reply:
`:0` means for `net.Listen` that it should listen for each network interface. In this case, we just using localhost.

Reply:
ok